cmd/adapter/adapter.go

```diff
@@ -0,0 +1,430 @@
+	}
+}
+
+//===================== adapter.cron =====================
+
+func (ad *Adapter) runCronQueries(ctx context.Context) {
+	if ad.cfg == nil || ad.cfg.Logrange == nil {
+		ad.logger.Warn("No config or Logrange config, skipping running cron queries")
+		return
+	}
+
+	ad.logger.Info("Running ", len(ad.cfg.Logrange.CronQueries), " cron queries")
+	for _, cq := range ad.cfg.Logrange.CronQueries {
+		ad.waitWg.Add(1)
+		go func(cq cronQuery) {
+			defer ad.waitWg.Done()
+			ad.runCronQuery(ctx, cq)
+		}(cq)
+	}
+}
+
+func (ad *Adapter) runCronQuery(ctx context.Context, cq cronQuery) {
+	ad.logger.Info("Entering into the loop to run \"", cq.Query, "\" every ", cq.PeriodSec, " seconds.")
+	for {
+		sleepDelay := time.Second * time.Duration(cq.PeriodSec)
+		res, err := ad.lCli.Execute(ctx, api.ExecRequest{Query: cq.Query})
+		if err != nil {
+			ad.logger.Warn("Could not connect to the server to run ", cq)
+			sleepDelay = time.Second
+		}
+
+		if res.Err != nil {
+			ad.logger.Error("Server returned error on executing ", cq.Query, " the err=", res.Err)
+		}
+
+		select {
+		case <-ctx.Done():
+			ad.logger.Info("Breaking up the runCronQuery loop for ", cq, " the context is closed.")
+			return
+		case <-time.After(sleepDelay):
+		}
+	}
+}
+
+//===================== adapter.sync =====================
+
+func (ad *Adapter) runSync(ctx context.Context) {
+	ad.logger.Info("Running sync every ", ad.cfg.SyncIntervalSec, " seconds...")
+	ticker := time.NewTicker(time.Second *
+		time.Duration(ad.cfg.SyncIntervalSec))
+
+	ad.waitWg.Add(1)
+	go func() {
+		for utils.Wait(ctx, ticker) {
+			ad.sync(ctx)
+		}
+		ad.logger.Warn("Sync stopped.")
+		ad.waitWg.Done()
+	}()
+}
+
+func (ad *Adapter) sync(ctx context.Context) {
+	ad.logger.Debug("sync(): Getting Logrange forwarder config...")
+	lrFwdCfg, err := ad.kCli.getLrFwdCfg()
+	if err != nil {
+		ad.logger.Error("sync(): Err=", err)
+		return
+	}
+
+	ad.logger.Debug("sync(): Getting Gravity forwarder config...")
+	grFwdCfgs, err := ad.kCli.getGrFwdCfg()
+	if err != nil {
+		ad.logger.Error("sync(): Err=", err)
+		return
+	}
+
+	ad.logger.Debug("sync(): Merging forwarder configs...")
+	newFwdCfg, err := ad.mergeFwdConfigs(lrFwdCfg.Forwarder, grFwdCfgs, ad.wTmpl)
+	if err != nil {
+		ad.logger.Error("sync(): Err=", err)
+		return
+	}
+
+	ad.logger.Info("sync(): Updating Logrange forwarder config: from=", lrFwdCfg.Forwarder, " to=", newFwdCfg)
+	lrFwdCfg.Forwarder = newFwdCfg
+	if err = ad.kCli.updateLrFwdCfg(lrFwdCfg); err != nil {
+		ad.logger.Error("sync(): Err=", err)
+	}
+}
+
+func (ad *Adapter) mergeFwdConfigs(lrFwdCfg *forwarder.Config, grFwdCfgs []*grForwarderCfg,
+	wCfgTmpl *forwarder.WorkerConfig) (*forwarder.Config, error) {
+
+	lrFwdCfg = deepcopy.Copy(lrFwdCfg).(*forwarder.Config)
+	lrFwdCfg.Workers = make([]*forwarder.WorkerConfig, 0, len(grFwdCfgs))
+
+	for _, grCfg := range grFwdCfgs {
+		wCfg := deepcopy.Copy(wCfgTmpl).(*forwarder.WorkerConfig)
+		if grCfg.Metadata.Name == "" {
```

Comment:
There should a separate validation function to keep this function's responsibility of only merging.
If I were implementing this, I would have `mergeForwarderConfig` either take a single gravity configuration or filter the invalid gravity configurations before calling this function.

Reply:
refactored, added filter function, applied (not yet merged to this branch)